feat: enable opt tool to parse files into LLVMAndRiscV dialect [1/2]

salinhkuhn opened the pull request:
This PR adds a new file to the LLVMAndRiscV project folder, which defines the infrastructure for parsing a region from an input file as a Com in the LLVMAndRiscV dialect. In future PRs, we will add functions and flags that enable interaction with the opt tool in the hybrid dialect setting therefore the suffix [1/2].
This PR also adds an import to the Opt.lean file to allow the opt tool to access the implemented extensions for the LLVMAndRiscV dialect to then add the corresponding flags within the opt tool next.

Commit:
feat: opt for hybrid

tobiasgrosser commented on SSA/Projects/InstCombine/LLVM/Opt.lean:
```diff
@@ -1,5 +1,7 @@
 import SSA.Projects.InstCombine.LLVM.Parser
 import Cli
+-- The import below is needed to access transformation flags for the LLVMAndRiscVhybrid dialect.
+import SSA.Projects.LLVMRiscV.ForOpt
```
I don't think the comment is needed here.

salinhkuhn replied:
removed it.

tobiasgrosser commented on SSA/Projects/LLVMRiscV/ForOpt.lean:
```diff
@@ -0,0 +1,35 @@
+import Cli
```
Instead of calling this file `ForOpt.lean`, can you name it based on what it offers, e.g., `ParseAndTransform.lean`. Or sth similar. There may be other tools that may want to use the same functions, so naming based on the features offered in afile is likely a good idea.

salinhkuhn replied:
done

tobiasgrosser commented on SSA/Projects/LLVMRiscV/ForOpt.lean:
```diff
@@ -0,0 +1,35 @@
+transformations in the hybrid dialect will be modeled
+as function calls from Opt.lean to functions defined in this
+file. This file can be seen as an extension of the Opt tool,
+specific to the LLVMAndRiscV dialect.
```
Your comments are now notably shorter than 80 characters. Is this intentional?

alexkeizer replied:
Beautiful documentation!

tobiasgrosser committed:
chore: update to lean nightly 2025-05-19 (#1243)

automatic update of mathlib + lean via GitHub action.

Co-authored-by: github-merge-queue <118344674+github-merge-queue@users.noreply.github.com>

alexkeizer commented on SSA/Projects/LLVMRiscV/ForOpt.lean:
```diff
@@ -0,0 +1,35 @@
+
+open MLIR AST InstCombine
+open LLVMRiscV
+/-
```
```suggestion
/!-
```

alexkeizer commented on SSA/Projects/LLVMRiscV/ForOpt.lean:
```diff
@@ -0,0 +1,35 @@
+    let res := mkCom (d:= LLVMPlusRiscV) region
+    match res with
+      | Except.error e => Except.error s!"Error:\n{reprStr e}"
+      | Except.ok res => Except.ok res
```
Check your indentation: 
```suggestion
    (Σ (Γ' : Ctxt LLVMPlusRiscV.Ty ) (eff : EffectKind)
      (ty : LLVMPlusRiscV.Ty ), Com LLVMPlusRiscV Γ' eff ty) :=
  let res := mkCom (d:= LLVMPlusRiscV) region
  match res with
    | Except.error e => Except.error s!"Error:\n{reprStr e}"
    | Except.ok res => Except.ok res
```

alexkeizer commented on SSA/Projects/LLVMRiscV/ForOpt.lean:
```diff
@@ -0,0 +1,35 @@
+def parseComFromFile_LLVMRiscV(fileName : String) :
+ IO (Option (Σ (Γ' :  Ctxt LLVMPlusRiscV.Ty ) (eff : EffectKind)
+  (ty :  LLVMPlusRiscV.Ty), Com LLVMPlusRiscV Γ' eff ty)) := do
+ parseRegionFromFile fileName regionTransform_LLVMRiscV
```
```suggestion
def parseComFromFile_LLVMRiscV (fileName : String) :
 IO (Option (Σ (Γ' :  Ctxt LLVMPlusRiscV.Ty ) (eff : EffectKind)
  (ty :  LLVMPlusRiscV.Ty), Com LLVMPlusRiscV Γ' eff ty)) := do
 parseRegionFromFile fileName regionTransform_LLVMRiscV
```
Also check your indentation here

alexkeizer reviewed (approved):
Please fix the indentation levels, then feel free to merge!

salinhkuhn committed:
feat: modify toString instance of LLVMAndRiscv (#1241)

This PR modifies the existing toString instance of the LLVMAndRiscV
dialect to no longer use the Repr instance internally. This allows
toString to differ from Repr.

---------

Co-authored-by: Sarah Kuhn <kuhnsa@student.ethz.ch>

salinhkuhn committed:
fix: add toString for LLVMAndRiscV.Ty (#1244)

This PR adds a toString instance for the LLVMAndRiscV.Ty.

---------

Co-authored-by: Sarah Kuhn <kuhnsa@student.ethz.ch>

salinhkuhn committed:
Update SSA/Projects/LLVMRiscV/ForOpt.lean

Co-authored-by: Alex Keizer <alex@keizer.dev>

salinhkuhn committed:
Update SSA/Projects/LLVMRiscV/ForOpt.lean

Co-authored-by: Alex Keizer <alex@keizer.dev>

Commit:
changed name of ForOpt to ParseAndTransform

Commit:
fix: rename to ParseAndTransform, fix comments

Commit:
fix: identation

Commit:
Merge remote-tracking branch 'origin/main' into opt-for-hybrid

Commit:
fix: identation
